crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -4511,26 +4524,36 @@ impl<'db> TypeInferenceBuilder<'db> {
                         _ => false,
                     };
 
-                    if bound_on_instance {
-                        self.context.report_lint(
-                            &UNRESOLVED_ATTRIBUTE,
-                            attribute,
-                            format_args!(
-                                "Attribute `{}` can only be accessed on instances, not on the class object `{}` itself.",
-                                attr.id,
-                                value_type.display(db)
-                            ),
-                        );
-                    } else {
-                        self.context.report_lint(
-                            &UNRESOLVED_ATTRIBUTE,
-                            attribute,
-                            format_args!(
-                                "Type `{}` has no attribute `{}`",
-                                value_type.display(db),
-                                attr.id
-                            ),
+                    let report_unresolved_attribute = self
+                        .index
+                        .is_expression_reachable(
+                            db,
+                            self.scope().file_scope_id(db),
+                            self.enclosing_expression_id(attribute),
                         );
+
+                    if report_unresolved_attribute {
+                        if bound_on_instance {
```

Comment:
it looks like computing `bound_on_instance` might be somewhat expensive, but we're doing it eagerly at the top of this branch whether or not `report_unresolved_attribute` resolves to `true`. Maybe turn that block of code into a lazy closure, so that we only evaluate it if we actually need to?

Reply:
Thanks. It's enough to just move that definition into the `if report_unresolved_attribute` branch.